File: app/schemas/market.py
"""
app.schemas.market — 行情数据模型（补充详情页）
"""
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


# ============== 行情 ==============
class QuoteItem(BaseModel):
    """单只股票行情"""
    symbol: str
    name: str
    price: float
    open: float
    high: float
    low: float
    close: float
    change: float
    change_pct: float
    volume: int
    amount: float
    turnover_rate: Optional[float] = None
    pe_ratio: Optional[float] = None
    pb_ratio: Optional[float] = None
    market_cap: Optional[float] = None
    circul_cap: Optional[float] = None


class QuoteResponse(BaseModel):
    """行情列表响应"""
    success: bool
    data: List[QuoteItem] = Field(default_factory=list)
    meta: dict = Field(default_factory=dict)


# ============== K 线 ==============
class KLineItem(BaseModel):
    """K 线数据"""
    date: str
    open: float
    close: float
    high: float
    low: float
    volume: int
    amount: float = 0.0
    amplitude: Optional[float] = None
    change_pct: Optional[float] = None
    change: Optional[float] = None
    turnover_rate: Optional[float] = None


class KLineResponse(BaseModel):
    """K线响应"""
    success: bool
    data: List[KLineItem] = Field(default_factory=list)
    meta: dict = Field(default_factory=dict)


# ============== 详情页（新增） ==============
class StockDetail(BaseModel):
    """股票详情（实时行情 + K 线 + 技术指标）"""
    symbol: str
    name: str
    market: str

    # 实时行情
    price: float
    open: float
    high: float
    low: float
    prev_close: float
    change: float
    change_pct: float
    volume: int
    amount: float

    # 财务指标
    turnover_rate: Optional[float] = None
    pe_ratio: Optional[float] = None
    pb_ratio: Optional[float] = None
    market_cap: Optional[float] = None
    circul_cap: Optional[float] = None

    # 技术指标（从 K 线计算）
    ma5: Optional[float] = None
    ma10: Optional[float] = None
    ma20: Optional[float] = None
    ma60: Optional[float] = None
    ma5_pct: Optional[float] = None  # 股价相对 MA5 偏离 %
    ma20_pct: Optional[float] = None  # 股价相对 MA20 偏离 %

    # 进阶技术指标
    rsi_6: Optional[float] = None
    rsi_14: Optional[float] = None
    macd_value: Optional[float] = None
    macd_signal: Optional[float] = None
    macd_hist: Optional[float] = None
    boll_upper: Optional[float] = None
    boll_mid: Optional[float] = None
    boll_lower: Optional[float] = None

    # K 线数据（精简，最近 60 个交易日）
    klines_daily: List[KLineItem] = Field(default_factory=list)
    klines_weekly: List[KLineItem] = Field(default_factory=list)


class StockDetailResponse(BaseModel):
    """股票详情响应"""
    success: bool
    data: Optional[StockDetail] = None
    message: str = ""

File: frontend/src/api/market.ts
/**
 * 行情 API 客户端
 * 对接后端 /api/v1/market/*
 */
import { request } from '@/utils/request'

export interface QuoteItem {
  symbol: string        // 股票代码，如 "600519.SH"
  name: string         // 股票名称，如 "贵州茅台"
  price: number        // 最新价
  change: number       // 涨跌额
  change_pct: number   // 涨跌幅 (%)
  prev_close: number   // 昨收
  open: number         // 开盘价
  high: number         // 最高价
  low: number          // 最低价
  volume: number       // 成交量（手）
  amount: number       // 成交额（元）
  market: 'A' | 'HK'  // 市场：A=A股，HK=港股
  update_time: string  // 更新时间 ISO
}

export interface MarketKLine {
  symbol: string
  period: '1d' | '1w' | '1m'
  data: Array<{
    time: string
    open: number
    high: number
    low: number
    close: number
    volume: number
  }>
}

/** 后端行情 API 通用响应包装 */
interface MarketApiResponse<T> {
  success: boolean
  data: T
  note?: string
}

/**
 * 将后端原始行情字段标准化为 QuoteItem
 * 后端字段: symbol, name, price, open, high, low, close(昨收), change, change_pct, volume, amount
 */
function normalizeQuote(raw: Record<string, any>): QuoteItem {
  return {
    symbol: raw.symbol || '',
    name: raw.name || raw.symbol || '',
    price: raw.price ?? 0,
    change: raw.change ?? 0,
    change_pct: raw.change_pct ?? 0,
    prev_close: raw.close ?? raw.prev_close ?? 0,  // 后端 close 字段即为昨收价
    open: raw.open ?? 0,
    high: raw.high ?? 0,
    low: raw.low ?? 0,
    volume: raw.volume ?? 0,
    amount: raw.amount ?? 0,
    market: raw.market || (String(raw.symbol || '').endsWith('.HK') ? 'HK' : 'A'),
    update_time: raw.update_time || new Date().toISOString(),
  }
}

/** 获取行情列表（全部或指定股票） */
export async function fetchQuotes(symbols?: string[]): Promise<QuoteItem[]> {
  const params: Record<string, string> = {}
  if (symbols && symbols.length > 0) {
    params.symbols = symbols.join(',')
  }
  // 后端返回 { success, data: [...] }，request 直接返回响应体
  const response = await request<MarketApiResponse<any[]>>('/market/quotes', {
    method: 'GET',
    params,
  })
  // 解包 data 数组并标准化每条数据
  const list = Array.isArray(response?.data) ? response.data : []
  return list.map(normalizeQuote)
}

/** 获取 K 线数据
 *  后端路由: GET /api/v1/market/kline/{symbol}?period=daily&count=100
 *  映射: date→time, period=daily→1d
 */
export async function fetchKLine(
  symbol: string,
  period: '1d' | '1w' | '1m' = '1d',
  limit = 100
): Promise<MarketKLine> {
  // period 映射: 前端 1d/1w/1m → 后端 daily/weekly/monthly
  const periodMap: Record<string, string> = { '1d': 'daily', '1w': 'weekly', '1m': 'monthly' }
  const response = await request<{ success: boolean; data: any[]; symbol: string }>(
    `/market/kline/${symbol}`,
    {
      method: 'GET',
      params: { period: periodMap[period] || 'daily', count: limit },
    }
  )
  return {
    symbol: response?.symbol || symbol,
    period,
    data: Array.isArray(response?.data)
      ? response.data.map((r: any) => ({
          time: r.date || r.time || '',
          open: r.open ?? 0,
          high: r.high ?? 0,
          low: r.low ?? 0,
          close: r.close ?? 0,
          volume: r.volume ?? 0,
        }))
      : [],
  }
}

// ──────────────────────────────────────────────
// T-M009 行情详情 API
// ──────────────────────────────────────────────

export interface KLineItem {
  date: string
  open: number
  close: number
  high: number
  low: number
  volume: number
}

export interface StockDetail {
  symbol: string
  name: string
  market: string
  price: number
  change: number
  change_pct: number
  open: number
  high: number
  low: number
  prev_close: number
  volume: number
  amount: number
  turnover_rate: number | null
  pe_ratio: number | null
  pb_ratio: number | null
  market_cap: number | null
  circul_cap: number | null
  ma5: number | null
  ma10: number | null
  ma20: number | null
  ma60: number | null
  ma5_pct: number | null
  ma20_pct: number | null
  rsi_6: number | null
  rsi_14: number | null
  macd_value: number | null
  macd_signal: number | null
  macd_hist: number | null
  boll_upper: number | null
  boll_mid: number | null
  boll_lower: number | null
  klines_daily: KLineItem[]
  klines_weekly: KLineItem[]
}

/** 实时行情快照（对接 detail 页面） */
export interface QuoteSnapshot {
  symbol: string
  name: string
  market: string
  price: number
  change: number
  change_pct: number
  open: number
  high: number
  low: number
  prev_close: number
  volume: number
  amount: number
  turnover_rate: number | null
  pe_ratio: number | null
  pb_ratio: number | null
  market_cap: number | null
  update_time: string
}

/** K线数据点（对接 detail 页面 KlineChart） */
export interface KlinePoint {
  date: string
  open: number
  close: number
  high: number
  low: number
  volume: number
  amount?: number
  change_pct?: number
}

/** 获取单股实时行情 */
export async function fetchQuote(symbol: string): Promise<QuoteSnapshot> {
  const detail = await marketApi.getDetail(symbol)
  const d = detail?.data || detail || {}
  return {
    symbol: d.symbol || symbol,
    name: d.name || symbol,
    market: d.market || 'A',
    price: d.price ?? 0,
    change: d.change ?? 0,
    change_pct: d.change_pct ?? 0,
    open: d.open ?? 0,
    high: d.high ?? 0,
    low: d.low ?? 0,
    prev_close: d.prev_close ?? 0,
    volume: d.volume ?? 0,
    amount: d.amount ?? 0,
    turnover_rate: d.turnover_rate ?? null,
    pe_ratio: d.pe_ratio ?? null,
    pb_ratio: d.pb_ratio ?? null,
    market_cap: d.market_cap ?? null,
    update_time: d.update_time || new Date().toISOString(),
  }
}

/** 获取K线（对接 detail 页面 KlineChart） */
export async function fetchKline(params: {
  symbol: string
  period?: 'day' | 'week' | 'month'
  count?: number
}): Promise<{ symbol: string; name: string; period: string; points: KlinePoint[] }> {
  const { symbol, period = 'day', count = 60 } = params
  const periodMap: Record<string, string> = { day: 'daily', week: 'weekly', month: 'monthly' }
  const res = await request<any>(`/market/kline/${symbol}`, {
    method: 'GET',
    params: { period: periodMap[period] || 'daily', count },
  })
  const data = res?.data || res || {}
  const points: KlinePoint[] = (Array.isArray(data) ? data : data.data || [])
    .map((r: any) => ({
      date: r.date || r.time || '',
      open: r.open ?? 0,
      close: r.close ?? 0,
      high: r.high ?? 0,
      low: r.low ?? 0,
      volume: r.volume ?? 0,
      amount: r.amount,
      change_pct: r.change_pct,
    }))
  return {
    symbol: data.symbol || symbol,
    name: data.name || symbol,
    period,
    points,
  }
}

export const marketApi = {
  getDetail: (symbol: string) => request<any>(`/market/detail/${symbol}`),
}
